File: 백준/Silver/9017. 크로스 컨트리/크로스 컨트리.java
import java.util.*;
import java.lang.*;
import java.io.*;

// The main method must be in a class named "Main".
class Main {
     public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        int N = Integer.parseInt(br.readLine());

        for (int i = 0; i < N; i++) {
            int trash = Integer.parseInt(br.readLine());
            String[] Nlist = br.readLine().split(" ");
            HashMap<Integer, Integer> moresixHash = new HashMap<>();

            // 팀별 주자 수 세기
            for (int j = 0; j < trash; j++) {
                int team = Integer.parseInt(Nlist[j]);
                moresixHash.put(team, moresixHash.getOrDefault(team, 0) + 1);
            }

            List<Integer> nolist = new ArrayList<>();
            // 6명 미만인 팀 구하기
            for (Integer intval : moresixHash.keySet()) {
                if (moresixHash.get(intval) < 6) {
                    nolist.add(intval);
                }
            }

            HashMap<Integer, int[]> NHash = new HashMap<>();
            int nowrank = 1;

            // 점수 계산
            for (String s : Nlist) {
                int nextteam = Integer.parseInt(s);
                if (!nolist.contains(nextteam)) {
                    if (NHash.containsKey(nextteam)) {
                        int[] data = NHash.get(nextteam);
                        if (data[0] < 4) {
                            data[0] += 1;
                            data[1] += nowrank;
                        } else if (data[0] == 4) {
                            data[0] += 1;
                            data[2] = nowrank;
                        }
                        NHash.put(nextteam, data); // 변경된 배열 값 업데이트
                    } else {
                        NHash.put(nextteam, new int[]{1, nowrank, 0}); // 새 팀 추가
                    }
                    nowrank += 1;
                }
            }

            // 최적의 팀 찾기
            int bestTeam = -1;
            int minSum = Integer.MAX_VALUE;
            int minFifth = Integer.MAX_VALUE;

            // 모든 팀을 순회하며 최적의 팀을 결정
            for (Integer key : NHash.keySet()) {
                int[] tmp = NHash.get(key);
                int sum = tmp[1];
                int fifth = tmp[2];
                
                if (sum < minSum || (sum == minSum && fifth < minFifth)) {
                    minSum = sum;
                    minFifth = fifth;
                    bestTeam = key;
                }
            }

            bw.write(bestTeam + "\n");
        }

        bw.flush();
        bw.close();
        br.close(); // BufferedReader 닫기
    }
}

File: 백준/Silver/9017. 크로스 컨트리/크로스 컨트리.py
import sys

N = int(input())

while N:

    trash = int(input())
    Nlist = list(map(int,sys.stdin.readline().split()))
    moresixHash = {}


    #팀별 수 세야지 .
    for nextval in Nlist:
        if nextval in moresixHash :
            moresixHash[nextval] +=1 
        else :
            moresixHash[nextval] = 1

 
    
    nolist = []
    #제외되는 팀 세기
    for i in moresixHash.keys() :
        if moresixHash[i] < 6 :
            nolist.append(i)

    NHash = {}
    nowrank = 1
    for nextteam in Nlist:
        if nextteam not in nolist:
            if nextteam in NHash:
                if NHash[nextteam][0] <4 :
                    NHash[nextteam][0] += 1
                    NHash[nextteam][1] += nowrank
                elif NHash[nextteam][0] == 4 :
                    NHash[nextteam][0] += 1
                    NHash[nextteam][2] = nowrank
            else :
                NHash[nextteam] = [1,nowrank,0]
            nowrank+=1

    
    sorted_teams = sorted(NHash.items(), key=lambda x: (x[1][1], x[1][2]))

    print(sorted_teams[0][0])

    N -= 1 
        